src/foam/nanos/notification/push/APNSPushService.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * @license
+ * Copyright 2024 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.notification.push',
+  name: 'APNSPushService',
+
+  documentation: `Not called directly, called through WebPushService 
+  TODO: not very msp friendly at the moment, there is no way for WebPushService to find the right apnsPushService for a given app`,
+
+  implements: [
+    'foam.nanos.notification.push.PushService',
+    'foam.nanos.auth.ServiceProviderAware'
+  ],
+
+  javaImports: [
+    'foam.core.X',
+    'foam.dao.*',
+    'foam.dao.ArraySink',
+    'foam.nanos.auth.*',
+    'foam.nanos.logger.Loggers',
+    'foam.mlang.MLang',
+    'foam.util.SafetyUtil',
+
+    'java.security.Security',
+    'java.util.List',
+    'java.util.HashMap',
+    'java.io.IOException',
+    'java.util.concurrent.ExecutionException',
+
+    'com.eatthepath.pushy.apns.*',
+    'com.eatthepath.pushy.apns.util.*',
+    'com.eatthepath.pushy.apns.util.concurrent.*',
+
+    'foam.nanos.security.KeyStoreManager',
+    'java.security.cert.X509Certificate',
+
+    'java.time.temporal.ChronoUnit',
+    'java.time.*'
+  ],
+
+  properties: [
+    {
+      class: 'Object',
+      of: 'com.eatthepath.pushy.apns.ApnsClient',
+      name: 'apnsClient',
+      transient: true,
+      javaFactory: `
+        return buildClient();
+      `
+    },
+    {
+      class: 'String',
+      name: 'apnsCredentialId',
+      documentation: 'credential id to find for host and key data'
+    },
+    {
+      class: 'Object',
+      of: 'foam.nanos.notification.push.APNSCredential',
+      name: 'apnsCredential',
+      transient: true,
+      javaFactory: `
+        return getCredentials(getX());
+      `
+    }
+  ],
+
+  methods: [
+    {
+      name: 'sendPushById',
+      javaCode:
+      `
+        //NO-OP
+        return false;
+      `
+    },
+    {
+      name: 'sendPush',
+      javaCode:
+      `
+        //NO-OP
+        return false;
+      `
+    },
+    {
+      name: 'buildClient',
+      type: 'ApnsClient',
+      throws: 'IOException',
+      javaCode:
+      `
+        // Implement in a refinement, based on certificate and pub priv key location
+      `
```

Comment:
it shouldn't compile unless it's an abstract method or returns an object or throw exception, should it?